### PR TITLE
[Stable] Introduce multiunifier

### PR DESCRIPTION
**Pull request by kasper-piskorski**

### CODE REVIEW

**Comment by burukuru** [comment]
!rtg

**Comment by burukuru** [comment]
!rtg

**Comment by burukuru** [comment]
!rtg

File: grakn-core/src/main/java/ai/grakn/graql/admin/MultiUnifier.java [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+    Unifier getUnifier();
+
+    /**
+     * @return the set of unifiers corresponding to this multiunifier
+     */
+    @CheckReturnValue
+    ImmutableSet<Unifier> unifiers();
+
+    /**
+     * ultiunifier merging by simple mapping addition (no variable clashes assumed)
```
Comment (fppt) [inline grakn-core/src/main/java/ai/grakn/graql/admin/MultiUnifier.java]: `multiunifier merging . . . `

File: grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationshipAtom.java [hunk 2]
```diff
@@ -761,73 +726,68 @@ private RelationshipAtom inferRoles(Answer sub){
...
+        return Sets.cartesianProduct(compatibleMappingsPerParentRP).stream()
+                .filter(list -> !list.isEmpty())
+                //check the same child rp is not mapped to multiple parent rps
+                .filter(list -> {
+                    List<RelationPlayer> listChildRps = list.stream().map(Pair::getKey).collect(Collectors.toList());
+                    //NB: this preserves cardinality instead of removing all occuring instances which is what we want
+                    return ReasonerUtils.subtract(listChildRps, this.getRelationPlayers()).isEmpty();
+                })
+                //check all parent rps mapped
+                .filter(list -> list.stream().map(Pair::getValue).collect(Collectors.toList()).containsAll(parentAtom.getRelationPlayers()))
```
Comment (fppt) [inline grakn-graql/src/main/java/ai/grakn/graql/internal/reasoner/atom/binary/RelationshipAtom.java]: Can you break this down better? This filter step is very hard to follow. Even just creating a method for it would help.

**Comment by kasper-piskorski** [comment]
Fixes issues with not returning all answers in some specific cases.

**Comment by kasper-piskorski** [comment]
!rtg - ai.grakn.factory.GraknTxJanusTest.ai.grakn.factory.GraknTxJanusTest
